Don't decompose functional composite ops in export inference IR

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #129547
* #129249
* #129092
* __->__ #128077

Recently we decided to split export IR into two different IRs (training vs inference). In the inference IR, one major change we decided to introduce was we wanted to keep the composite ops that user specified in the IR. This PR does that by overriding the CompositeImplicitAutograd decomp in export inference path.

Differential Revision: [D58701607](https://our.internmc.facebook.com/intern/diff/D58701607)

tugsbayasgalan committed:
Attempt at making linear not decompose

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Attempt at making linear not decompose"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Attempt at making linear not decompose"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Attempt at making linear not decompose"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Attempt at making linear not decompose"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Don't decompose functional composite ops in export inference IR"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Don't decompose functional composite ops in export inference IR"


Recently we decided to split export IR into two different IRs (training vs inference). In the inference IR, one major change we decided to introduce was we wanted to keep the composite ops in the IR by default. This PR does that by overriding the CompositeImplicitAutograd decomp in export inference path. 


[ghstack-poisoned]

bdhirsh commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -331,15 +331,16 @@ def can_handle_tensor(x):
         return r
 
     # For pre-autograd tracing, we do not want to run CompositeImplicit decomps.
-    if not pre_dispatch and func not in [
```
Why are we removing this check? (if pre_dispatch is False and `export_inference` is False, we will now decompose?

bdhirsh commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -331,15 +331,16 @@ def can_handle_tensor(x):
-            r = func.decompose(*args, **kwargs)
-            if r is not NotImplemented:
-                return r
+            if not proxy_mode.export_inference:
```
cc @ezyang / @zou3519, any thoughts on another flag gunking up ProxyMode?

The use case is that Tugsuu wants to avoid decomposing CompositeImplicit ops in the inference code path for export. Which requires auditing all the places we normally decompose:

(1) autograd key (CompositeImplicits)
(2) FunctionalTensorMode
(3) ProxyMode

The alternative we discussed was having a layer that converts all composite implicit ops into custom ops during tracing, then a graph pass that converts back after. The main annoyance with that approach though is we'd need to add infra to interpose and actually turn all composite implicits into custom ops during tracing (probably in pre-dispatch somewhere).

ezyang replied:
I'd prefer the flag be named after what it does, rather than who is using it, but being able to impede func.decompose calls seems like a reasonable knob to have.

tugsbayasgalan replied:
It seems the custom op approach won't be as easy because since we are doing it at pre-dispatch level, we need to make it work with subclass modes in the future.

zou3519 replied:
+1 for Ed's comment. A knob on func.decompose seems reasonable and the name should reflect that.

One alternative to threading the flag around these modes is to have it as a single config option somewhere

tugsbayasgalan committed:
Update on "Don't decompose functional composite ops in export inference IR"


Recently we decided to split export IR into two different IRs (training vs inference). In the inference IR, one major change we decided to introduce was we wanted to keep the composite ops in the IR by default. This PR does that by overriding the CompositeImplicitAutograd decomp in export inference path. 


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Don't decompose functional composite ops in export inference IR"


Recently we decided to split export IR into two different IRs (training vs inference). In the inference IR, one major change we decided to introduce was we wanted to keep the composite ops in the IR by default. This PR does that by overriding the CompositeImplicitAutograd decomp in export inference path. 


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Don't decompose functional composite ops in export inference IR"


Recently we decided to split export IR into two different IRs (training vs inference). In the inference IR, one major change we decided to introduce was we wanted to keep the composite ops in the IR by default. This PR does that by overriding the CompositeImplicitAutograd decomp in export inference path. 


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Don't decompose functional composite ops in export inference IR"


Recently we decided to split export IR into two different IRs (training vs inference). In the inference IR, one major change we decided to introduce was we wanted to keep the composite ops in the IR by default. This PR does that by overriding the CompositeImplicitAutograd decomp in export inference path. 


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Don't decompose functional composite ops in export inference IR"


Recently we decided to split export IR into two different IRs (training vs inference). In the inference IR, one major change we decided to introduce was we wanted to keep the composite ops in the IR by default. This PR does that by overriding the CompositeImplicitAutograd decomp in export inference path. 


[ghstack-poisoned]

zhxchen17 reviewed (commented):
Can you log the preserved ops in log_export_usage(event="export.enter") ?

tugsbayasgalan committed:
Update on "Don't decompose functional composite ops in export inference IR"


Recently we decided to split export IR into two different IRs (training vs inference). In the inference IR, one major change we decided to introduce was we wanted to keep the composite ops that user specified in the IR. This PR does that by overriding the CompositeImplicitAutograd decomp in export inference path. 


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Don't decompose functional composite ops in export inference IR"


Recently we decided to split export IR into two different IRs (training vs inference). In the inference IR, one major change we decided to introduce was we wanted to keep the composite ops that user specified in the IR. This PR does that by overriding the CompositeImplicitAutograd decomp in export inference path. 


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Don't decompose functional composite ops in export inference IR"


Recently we decided to split export IR into two different IRs (training vs inference). In the inference IR, one major change we decided to introduce was we wanted to keep the composite ops that user specified in the IR. This PR does that by overriding the CompositeImplicitAutograd decomp in export inference path. 


[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58701607).

tugsbayasgalan committed:
Update on "Don't decompose functional composite ops in export inference IR"


Recently we decided to split export IR into two different IRs (training vs inference). In the inference IR, one major change we decided to introduce was we wanted to keep the composite ops that user specified in the IR. This PR does that by overriding the CompositeImplicitAutograd decomp in export inference path.

Differential Revision: [D58701607](https://our.internmc.facebook.com/intern/diff/D58701607)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Don't decompose functional composite ops in export inference IR"


Recently we decided to split export IR into two different IRs (training vs inference). In the inference IR, one major change we decided to introduce was we wanted to keep the composite ops that user specified in the IR. This PR does that by overriding the CompositeImplicitAutograd decomp in export inference path.

Differential Revision: [D58701607](https://our.internmc.facebook.com/intern/diff/D58701607)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Don't decompose functional composite ops in export inference IR"


Recently we decided to split export IR into two different IRs (training vs inference). In the inference IR, one major change we decided to introduce was we wanted to keep the composite ops that user specified in the IR. This PR does that by overriding the CompositeImplicitAutograd decomp in export inference path.

Differential Revision: [D58701607](https://our.internmc.facebook.com/intern/diff/D58701607)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Don't decompose functional composite ops in export inference IR"


Recently we decided to split export IR into two different IRs (training vs inference). In the inference IR, one major change we decided to introduce was we wanted to keep the composite ops that user specified in the IR. This PR does that by overriding the CompositeImplicitAutograd decomp in export inference path.

Differential Revision: [D58701607](https://our.internmc.facebook.com/intern/diff/D58701607)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Don't decompose functional composite ops in export inference IR"


Recently we decided to split export IR into two different IRs (training vs inference). In the inference IR, one major change we decided to introduce was we wanted to keep the composite ops that user specified in the IR. This PR does that by overriding the CompositeImplicitAutograd decomp in export inference path.

Differential Revision: [D58701607](https://our.internmc.facebook.com/intern/diff/D58701607)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Don't decompose functional composite ops in export inference IR"


Recently we decided to split export IR into two different IRs (training vs inference). In the inference IR, one major change we decided to introduce was we wanted to keep the composite ops that user specified in the IR. This PR does that by overriding the CompositeImplicitAutograd decomp in export inference path.

Differential Revision: [D58701607](https://our.internmc.facebook.com/intern/diff/D58701607)

[ghstack-poisoned]

bdhirsh commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -97,6 +97,15 @@ class FunctionalTensor(torch.Tensor):
         torch.ops.aten._batch_norm_impl_index.default,  # type: ignore[has-type]
         torch.ops.aten.cudnn_batch_norm.default,  # type: ignore[has-type]
         torch.ops.aten.miopen_batch_norm.default,  # type: ignore[has-type]
+        torch.ops.aten.atleast_1d.default,  # type: ignore[has-type]
```
how did you end up finding these?

tugsbayasgalan replied:
@pianpwk ran the schema check mode on pre-dispatch key stack

bdhirsh commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -331,41 +340,31 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+            if func in FunctionalTensor.maybe_aliasing_or_mutating_ops:
+                return True
+
+            # We unconditionally decompose ops that are mutating or aliasing ops
```
this comment and the one above it is kind of confusing, maybe worth elaborating:

(1) we unconditionally decompose maybe-aliasing or maybe-mutating ops, because we **must** know statically of an op mutates or aliasing in order to functionalize it properly

(2) for mutating ops that have CompositeImplicit decomps, we choose to decompose them today. In theory, we could walk this back and avoid decomposing them later if we need to.

bdhirsh commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -331,41 +340,31 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+
+            # If we are here, it means we are seeing functional composite op.
+            # For pre-dispatch IR or export inference IR, we wont' decompose them
+            if self.export or self.pre_dispatch:
```
dumb question: are these always both true? Or do we have a use case for one being false and the other true (if not, let's just assert that they're both true or both false somewhere to make it clear in the code)

tugsbayasgalan replied:
Yep for now, they are both True. I can assert it loudly.

bdhirsh replied:
yes please to the assert (probably in the __init__).

Can you either do it in this PR or add an explicit TODO

bdhirsh replied:
still looking for the todo!

bdhirsh commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -331,41 +340,31 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+            # For pre-dispatch IR or export inference IR, we wont' decompose them
+            if self.export or self.pre_dispatch:
+                if func.namespace not in ["aten", "prim"]:
+                    warnings.warn(
```
hmmm.... this warning doesn't feel very actionable

here's a tentative alternative proposal: if the only time we can end up here is if we hit a custom op, then:

(1) if the custom op is pt2-compliant (marked with the `Tag.pt2_compliant` tag), then it promises to be traceable so we should leave it in

(2) if the custom op is not pt2-compliant, then we should error loudly and tell the user to make their op compliant. If dynamo is running (via strict export) we should never end up here, since dynamo will have graph broken / error'd on the op.

tugsbayasgalan replied:
ok sounds good!

tugsbayasgalan replied:
Will do as follow up

bdhirsh commented on torch/export/_trace.py:
```diff
@@ -225,6 +247,104 @@ def convert_to_fake(x):
     return fake_args, fake_kwargs, fake_params_buffers, fake_mode
 
 
+def _skip_autograd_kernel(*args, **kwargs):
```
this can all be deleted too, right?

tugsbayasgalan replied:
Hmm this is still needed because this is the one that actually skips autograd.

zou3519 replied:
If you want this to be faithful, this needs to be a opoverload.redispatch

bdhirsh commented on torch/export/_trace.py:
```diff
@@ -225,6 +247,104 @@ def convert_to_fake(x):
+    patched_ops = set()
+    for op_overload in ops_to_preserve:
+
+        def can_preserve(op_overload):
```
this function probably deserves a long comment explaining how we decide when ops can be preserved in export.

bdhirsh commented on torch/export/_trace.py:
```diff
@@ -1465,20 +1590,27 @@ def _export(
-        _is_torch_jit_trace,
-    )
 
+    # TODO we only want to do this for aot-dispatch for now. Pre-dispatch will be deleted soon
```
wdym pre-dispatch will be deleted soon? 😛

tugsbayasgalan replied:
Yep should be more clear here! I meant we are deleting pre-dispathc functional export.

bdhirsh commented on torch/export/_trace.py:
```diff
@@ -1380,6 +1500,7 @@ def _export(
     _allow_complex_guards_as_runtime_asserts: bool = False,
     _disable_forced_specializations: Optional[bool] = False,
     _is_torch_jit_trace: bool = False,
+    _preserve_ops: Optional[List[OpOverload]] = None,
```
just curious, but did we finalize on "separate decomp table + preserve_ops list" being the final API we want for export?

tugsbayasgalan replied:
Yep i think so

bdhirsh commented on torch/export/_trace.py:
```diff
@@ -225,6 +247,104 @@ def convert_to_fake(x):
+        def can_preserve(op_overload):
+            if op_overload in FunctionalTensor.maybe_aliasing_or_mutating_ops:
+                return False
+            if op_overload in FunctionalTensor.metadata_fns:
```
hmmm actually - this `ops_to_preserve` list comes from the user, right?

Right now, it looks like you have a long `can_preserve` function that checks if we can **actually** preserve each of the ops that the user specified. But it seems wrong to silently ignore what the user asked - what do you think of erroring loudly if the user puts an op in the list that is invalid?

tugsbayasgalan replied:
Yeah good idea, i will just error

bdhirsh replied:
hmm I don't think I see the error anywhere?

bdhirsh commented on torch/export/_trace.py:
```diff
@@ -225,6 +247,104 @@ def convert_to_fake(x):
+        for override_dispatch_key in [
+            torch._C.DispatchKey.AutogradCPU,
+            torch._C.DispatchKey.AutogradCUDA,
+            torch._C.DispatchKey.AutogradMeta,
```
to be safe you probably want this to include all of the `AutogradBACKEND` keys, otherwise this will just be broken for other backends (e.g. `AutogradXPU`, `AutogradPrivateUse1`, etc)

tugsbayasgalan replied:
Hmm yeah it felt kinda ugly to list all of them here and i thought we could get away with just these three haha. Realisitcally, would we ever hit the other cases in torch.export?

bdhirsh replied:
(1) if we don't do it, and someone tries to export a model where their inputs are on a custom device, we'll get divergent behavior which feels bad

(2) the good news is that this list of keys is no longer growing (since new custom backends all need to use the `AutogradPrivateUse1` key.

So I would proably lean toward just hardcoding the list.

bdhirsh commented on torch/export/_trace.py:
```diff
@@ -225,6 +247,104 @@ def convert_to_fake(x):
+            if override_dispatch_key not in op_overload.py_kernels:
+                # conv1d, conv2d, conv3d don't work with default Autograd key
+                op_overload.py_impl(override_dispatch_key)(
+                    functools.partial(_skip_autograd_kernel, kernel=op_overload)
```
this should be updated to use the autograd_error kernel that custom ops uses, right? (fyi @zou3519)

tugsbayasgalan replied:
Oh yeah i forgot to do it, will do.

zou3519 replied:
@tugsbayasgalan and I talked offline, and we don't think this is necessary: what situations are we trying to error out on? If we are exporting a module, then hoping that `exported_program.module().forward(x).backward()` will error out, then the operator needs to use the autograd_error kernel during the execution of the forward (and not during export)

bdhirsh replied:
I agree that this won't help much if the user does the above. My thought was mostly that:

(1) "not decomposing a CompositeImplicit op" during tracing has safety issues if you happen to try to use the autograd engine during tracing (e.g. if you're tracing out the joint graph)

(2) if we're adding the capability to skip CompositeImplicit decomps for export now, we may as well add it in a safe way that will automatically error when someone eventually wants to get joint graph tracing to work (instead of having to remember the correctness conditions and figure out how to error properly later).

If this is easy to do (e.g. through the autograd_error API that custom ops uses), then it seems useful to do it right the first time. But lmk if you disagree / think this is a pain.

tugsbayasgalan replied:
I tried doing:
```
op_overload.py_impl(AutogradCPU)(make_autograd_impl(op_overload, Info(None, None)))
```
But ran into following error:
```
    flat_f_outs = f(*flat_f_args)
  File "/data/users/tmanlaibaatar/pytorch/torch/_functorch/_aot_autograd/utils.py", line 178, in flat_fn
    tree_out = fn(*args, **kwargs)
  File "/data/users/tmanlaibaatar/pytorch/torch/_functorch/_aot_autograd/traced_function_transforms.py", line 760, in functional_call
    out = PropagateUnbackedSymInts(mod).run(
  File "/data/users/tmanlaibaatar/pytorch/torch/fx/interpreter.py", line 146, in run
    self.env[node] = self.run_node(node)
  File "/data/users/tmanlaibaatar/pytorch/torch/fx/experimental/symbolic_shapes.py", line 5462, in run_node
    result = super().run_node(n)
  File "/data/users/tmanlaibaatar/pytorch/torch/fx/interpreter.py", line 203, in run_node
    return getattr(self, n.op)(n.target, args, kwargs)
  File "/data/users/tmanlaibaatar/pytorch/torch/fx/interpreter.py", line 275, in call_function
    return target(*args, **kwargs)
  File "/data/users/tmanlaibaatar/pytorch/torch/_ops.py", line 670, in __call__
    return self_._op(*args, **kwargs)
  File "/data/users/tmanlaibaatar/pytorch/torch/_library/autograd.py", line 113, in autograd_impl
    result = forward_no_grad(*args, Metadata(keyset, keyword_only_args))
  File "/data/users/tmanlaibaatar/pytorch/torch/_library/autograd.py", line 40, in forward_no_grad
    result = op.redispatch(keyset & _C._after_autograd_keyset, *args, **kwargs)
TypeError: unsupported operand type(s) for &: 'FunctionalTensor' and 'torch._C.DispatchKeySet'
```

It seems like i am using this API kinda wrong. When i was reading the code, it looks like we are handling the last arg little differently https://github.com/pytorch/pytorch/blob/d52684e9a8b95f8b0d06f0bf08e6a39846cb3ae6/torch/_library/autograd.py#L44

@zou3519 any clues?

zou3519 replied:
There's two calling conventions for the Dispatcher. If you have an operator "sin(Tensor x) -> Tensor" -- you can register a kernel that accepts "(x: Tensor)" or a kernel that accepts a DispatchKeySet first: "(keyset: DispatchKeySet, x: Tensor)".

To use `make_autograd_impl`, we need to add support for the latter calling convention in `py_impl`. This shouldn't be hard:
- we'd add an extra argument to `py_impl` that is a bool flag for the calling convention and then store this information somewhere
- then we need to call into the py_impl with the calling convention.

You need to do this anyways to make the redispatch faithful (that is, the redispatch should use opoverload.redispatch).

bdhirsh commented on torch/export/_trace.py:
```diff
@@ -225,6 +247,104 @@ def convert_to_fake(x):
+                    functools.partial(_skip_autograd_kernel, kernel=op_overload)
+                )
+        if torch._C.DispatchKey.CompositeImplicitAutograd in op_overload.py_kernels:
+            del op_overload.py_kernels[torch._C.DispatchKey.CompositeImplicitAutograd]
```
this doesn't feel very safe.

In your finally block, you probably want to undo all of the dispatch table changes you made so far (swapping out the Meta and CompositeImplicitAutograd registrations)

tugsbayasgalan replied:
Yep i save the table before this and set it afterwards.

bdhirsh reviewed (changes requested):
I think the high-level design looks fine, as long as it works! left some comments

tugsbayasgalan committed:
Update on "Don't decompose functional composite ops in export inference IR"


Recently we decided to split export IR into two different IRs (training vs inference). In the inference IR, one major change we decided to introduce was we wanted to keep the composite ops that user specified in the IR. This PR does that by overriding the CompositeImplicitAutograd decomp in export inference path.

Differential Revision: [D58701607](https://our.internmc.facebook.com/intern/diff/D58701607)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Don't decompose functional composite ops in export inference IR"


Recently we decided to split export IR into two different IRs (training vs inference). In the inference IR, one major change we decided to introduce was we wanted to keep the composite ops that user specified in the IR. This PR does that by overriding the CompositeImplicitAutograd decomp in export inference path.

Differential Revision: [D58701607](https://our.internmc.facebook.com/intern/diff/D58701607)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Don't decompose functional composite ops in export inference IR"


Recently we decided to split export IR into two different IRs (training vs inference). In the inference IR, one major change we decided to introduce was we wanted to keep the composite ops that user specified in the IR. This PR does that by overriding the CompositeImplicitAutograd decomp in export inference path.

Differential Revision: [D58701607](https://our.internmc.facebook.com/intern/diff/D58701607)

[ghstack-poisoned]

zou3519 commented on test/export/test_export.py:
```diff
@@ -64,7 +64,7 @@
 
     HAS_TORCHREC = True
-except ImportError:
+except (ImportError, AttributeError):
```
Why the change? Are we actually importing torchrec correctly in CI?

tugsbayasgalan replied:
Yeah sorry this is my local thing. Not needed.

zou3519 commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -97,6 +97,15 @@ class FunctionalTensor(torch.Tensor):
+        torch.ops.aten.atleast_2d.default,  # type: ignore[has-type]
+        torch.ops.aten.atleast_3d.default,  # type: ignore[has-type]
+        torch.ops.aten.cartesian_prod.default,  # type: ignore[has-type]
+        torch.ops.aten.conj_physical.default,  # type: ignore[has-type]
+        torch.ops.aten.alpha_dropout.default,  # type: ignore[has-type]
+        torch.ops.aten.feature_dropout.default,  # type: ignore[has-type]
+        torch.ops.aten.feature_alpha_dropout.default,  # type: ignore[has-type]
```
@tugsbayasgalan what's the plan with all of these? It seems like a silent incorrectness bug if an operator misrepresents its schema (unless it is one of the unsafe_* ops). If we agree that these are bugs, can you file some issues for them?

bdhirsh replied:
I think for any of these "maybe-aliasing/mutation ops", our options are something like:

(1) always decompose them, force users of export to live with it

(2) for maybe-aliasing ops: change the op to never alias via CoW (only for export, not eager)

(3) for maybe-aliasing ops: change the op to never alias for both export and eager

Right now...

*  export is doing (1) for all of these ops

* except for dropout export does (2) (and it carefully tries to error on programs whose semantics diverge due to the dropout behavior change from eager)

* and in theory we might do (3) for `reshape()`, if we are willing to sit through the warn / deprecation cycle? I'm not sure what the progress here has been though.

zou3519 commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -331,41 +344,35 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+                    # TODO (tmanlaibaatar) check if the op is PT2 compliant
+                    warnings.warn(
+                        f"At pre-dispatch tracing, we assume that any custom op marked with "
+                        f"CompositeImplicitAutograd and functional are safe to not decompose. "
```
What do you mean by "marked with functional"? That the schema looks functional?

zou3519 commented on torch/export/exported_program.py:
```diff
@@ -118,6 +120,130 @@ def _fx_collection_equivalence_fn(
+        # 1. The op should be known statically that it is functional
+        # 2. If it is maybe aliasing, we decompose because we must know if an op
+        #    is mutating or aliasing.
+        # TODO (tmanlaibaatar) make this utility function and share it with functional_tensor
```
Could you file an issue for this TODO (and any others in this PR) and assign yourself so we can better track them?

tugsbayasgalan replied:
Done!

zou3519 commented on torch/export/exported_program.py:
```diff
@@ -118,6 +120,130 @@ def _fx_collection_equivalence_fn(
+
+        saved_tables[op_overload] = op_overload.py_kernels.copy()
+        patched_ops.add(op_overload)
+        for override_dispatch_key in [
+            torch._C.DispatchKey.AutogradCPU,
+            torch._C.DispatchKey.AutogradCUDA,
+            torch._C.DispatchKey.AutogradMeta,
+        ]:
+            if override_dispatch_key not in op_overload.py_kernels:
+                # conv1d, conv2d, conv3d don't work with default Autograd key
+                op_overload.py_impl(override_dispatch_key)(
+                    functools.partial(_skip_autograd_kernel, kernel=op_overload)
+                )
```
Does it not work to override CompositeImplicitAutograd? If not then you probably want to override all Autograd* keys. Also, how does this work with inference_mode?

tugsbayasgalan replied:
Nope. Because C++ decomp from AutogradCPU etc from C++ is higher precedence than python CIA decomp

zou3519 reviewed (commented):
Looks reasonable. We probably want to make the autograd registration more legit (either now or later)

tugsbayasgalan commented on torch/export/exported_program.py:
```diff
@@ -118,6 +120,130 @@ def _fx_collection_equivalence_fn(
+            torch._C.DispatchKey.CompositeImplicitAutograd, *args, **kwargs
+        )
+    else:
+        return NotImplemented
```
I think we will probably never hit this case. Can just assert probably to make code look cleaner.

tugsbayasgalan commented on torch/export/exported_program.py:
```diff
@@ -526,7 +652,9 @@ def _num_lifted_params_buffers(self):
-        self, decomp_table: Optional[Dict[torch._ops.OperatorBase, Callable]] = None
+        self,
+        decomp_table: Optional[Dict[torch._ops.OperatorBase, Callable]] = None,
+        _preserve_ops: Optional[List[torch._ops.OpOverload]] = None,
```
Yep!

tugsbayasgalan commented on torch/export/exported_program.py:
```diff
@@ -554,6 +682,9 @@ def _get_placeholders(gm):
         if decomp_table is None:
             decomp_table = core_aten_decompositions()
 
+        if _preserve_ops is None:
```
Logging point here

bdhirsh commented on torch/export/exported_program.py:
```diff
@@ -118,6 +120,130 @@ def _fx_collection_equivalence_fn(
+        return result
+
+
+def _register_cia_to_meta(*args, **kwargs):
```
nit: signature here can just be `_register_cia_to_meta(*, kernel)` right?

bdhirsh commented on test/export/test_export.py:
```diff
@@ -661,6 +661,36 @@ def forward(self, x, weight, bias):
+        gm = torch.export._trace._export(
+            Foo(),
+            (torch.randn(3, 3),),
+            pre_dispatch=False,
+            _preserve_ops=[torch.ops.aten.linear.default],
```
+1, it would be nice to have a test that properly errors if you try to include any of:

(1) an op that is not CompositeImplicit
(2) a CompositeImplicit op that is "maybe-aliasing"
(3) a CompositeImplicit op that is mutating (since we will always decompose these in FunctionalTensorMode today)

tugsbayasgalan committed:
Update on "Don't decompose functional composite ops in export inference IR"


Recently we decided to split export IR into two different IRs (training vs inference). In the inference IR, one major change we decided to introduce was we wanted to keep the composite ops that user specified in the IR. This PR does that by overriding the CompositeImplicitAutograd decomp in export inference path.

Differential Revision: [D58701607](https://our.internmc.facebook.com/intern/diff/D58701607)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Don't decompose functional composite ops in export inference IR"


Recently we decided to split export IR into two different IRs (training vs inference). In the inference IR, one major change we decided to introduce was we wanted to keep the composite ops that user specified in the IR. This PR does that by overriding the CompositeImplicitAutograd decomp in export inference path.

Differential Revision: [D58701607](https://our.internmc.facebook.com/intern/diff/D58701607)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Don't decompose functional composite ops in export inference IR"


Recently we decided to split export IR into two different IRs (training vs inference). In the inference IR, one major change we decided to introduce was we wanted to keep the composite ops that user specified in the IR. This PR does that by overriding the CompositeImplicitAutograd decomp in export inference path.

Differential Revision: [D58701607](https://our.internmc.facebook.com/intern/diff/D58701607)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Don't decompose functional composite ops in export inference IR"


Recently we decided to split export IR into two different IRs (training vs inference). In the inference IR, one major change we decided to introduce was we wanted to keep the composite ops that user specified in the IR. This PR does that by overriding the CompositeImplicitAutograd decomp in export inference path.

Differential Revision: [D58701607](https://our.internmc.facebook.com/intern/diff/D58701607)

[ghstack-poisoned]

bdhirsh commented on test/export/testing.py:
```diff
@@ -2,6 +2,196 @@
+
+aten = torch.ops.aten
+
+COMPOSITE_OPS_THAT_CAN_BE_PRESERVED = [
```
hmm... why do we want a hardcoded list here for testing? I'm worried that people will read this and expect that it is a source of truth, but it will inevitable go stale if/when someone adds a new CompositeImplicit op.

bdhirsh commented on torch/export/exported_program.py:
```diff
@@ -118,6 +122,102 @@ def _fx_collection_equivalence_fn(
+            return True
+
+        if not can_preserve(op_overload):
+            # TODO (tmanlaibaatar) Better error message
```
you might just want to have the `can_preserve()` fn error directly so you can raise the different errors for the different cases, instead of having to reconstruct the different cases here.

tugsbayasgalan replied:
Yep sounds good!

bdhirsh commented on torch/export/exported_program.py:
```diff
@@ -118,6 +122,102 @@ def _fx_collection_equivalence_fn(
+        for override_dispatch_key in [
+            torch._C.DispatchKey.AutogradCPU,
+            torch._C.DispatchKey.AutogradCUDA,
+            torch._C.DispatchKey.AutogradMeta,
```
the other AutogradFoo dispatch keys are still missing here?

tugsbayasgalan replied:
Just added

tugsbayasgalan committed:
Update on "Don't decompose functional composite ops in export inference IR"


Recently we decided to split export IR into two different IRs (training vs inference). In the inference IR, one major change we decided to introduce was we wanted to keep the composite ops that user specified in the IR. This PR does that by overriding the CompositeImplicitAutograd decomp in export inference path.

Differential Revision: [D58701607](https://our.internmc.facebook.com/intern/diff/D58701607)

[ghstack-poisoned]

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -1015,6 +1047,106 @@ def forward(self, x, y):
+            _ = torch.export.export(
+                Foo(),
+                (torch.randn(3, 3),),
+            ).run_decompositions({}, _preserve_ops=(torch.ops.aten.chunk.default,))
```
Added

zou3519 reviewed (commented):
LGTM, leaving the final stamp for Brian

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58701607).

bdhirsh commented on torch/export/exported_program.py:
```diff
@@ -118,6 +122,127 @@ def _fx_collection_equivalence_fn(
+        #    is mutating or aliasing.
+        # TODO (tmanlaibaatar) make this utility function and share it with functional_tensor
+        # decomp part. (https://github.com/pytorch/pytorch/issues/129431)
+        def can_preserve(op_overload):
```
nit: this should probably be named something like `assert_valid_to_preserve`? (can_preserve implies true/false, but this whole function is now just an assertion)

bdhirsh commented on test/export/testing.py:
```diff
@@ -2,6 +2,196 @@
+
+aten = torch.ops.aten
+
+_COMPOSITE_OPS_THAT_CAN_BE_PRESERVED_TESTING_ONLY = [
```
i would still comment in the code that this list isn't required to be complete, but is convenient for getting "pretty good" test coverage or something (someone might look at it later and wonder if it's bad that an op is missing from the list)

bdhirsh commented on torch/export/exported_program.py:
```diff
@@ -118,6 +122,127 @@ def _fx_collection_equivalence_fn(
+
+            if not torch._C._dispatch_has_kernel(op_overload.name()):
+                raise RuntimeError(
+                    f"{op_overload} is a TorchScript op, we can't preserve it as is"
```
oh is this true? (that the above check returning false implies that it is a torchscript op). that doesn't seem obvious to me. If it's not, maybe just make the error message match the behavior ("op_overload is not an operator registered to the dispatcher" or something)

bdhirsh commented on torch/export/exported_program.py:
```diff
@@ -118,6 +122,127 @@ def _fx_collection_equivalence_fn(
+                )
+            if op_overload in FunctionalTensor.metadata_fns:
+                raise RuntimeError(
+                    f"{op_overload} is a metadata query function, "
```
nit: you might want to make these error messages all a bit more actionable to someone who runs into them: "please remove this op from `_preserve_ops` list"

bdhirsh reviewed (approved):
lgtm!

tugsbayasgalan committed:
Update on "Don't decompose functional composite ops in export inference IR"


Recently we decided to split export IR into two different IRs (training vs inference). In the inference IR, one major change we decided to introduce was we wanted to keep the composite ops that user specified in the IR. This PR does that by overriding the CompositeImplicitAutograd decomp in export inference path.

Differential Revision: [D58701607](https://our.internmc.facebook.com/intern/diff/D58701607)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58701607).

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58701607).

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58701607).
